### PR TITLE
fix: reverse one-to-one relationships

### PR DESCRIPTION
**Pull request by PagingMatt**
This fixes the user-reported issue where `ONE_TO_ONE` relationships would only be resolved in the forward direction as opposed to also in the reverse direction.

Consider a simple data structure of a Biography and an Author. they are linked by a `ONE_TO_ONE` relationship whereby Biography has an `author` field. Unfortunately in the Noloco backend right now there is no `reverseName` property added to this field, meaning our existing logic to lookup the Biography from an Author query would not work.

The required solution was to extend `find_relationship_data_type` so that when we are looking up reverse relationships we can handle the case where there is not a `reverseName` for `ONE_TO_ONE` relationships in which case we should just look for the relationship name to match the data type name and for the forwards relationship field on that data type to have a matching type.

This then broke how this code was used upstream as it previously expected that all reverse relationships would return collections rather than single results. The fix for this was to make `find_relationship_data_type` pass back whether the result of resolving this relationship is a collection or not.

---

I tested against my reproduction of the user issue and confirmed this has fixed their use case. Given this is a core part of code, a bit more thorough regression testing will be needed to feel comfortable pushing this up. The nice thing about `find_relationship_data_type` is it now encapsulated pretty much everything about stepping down into a relationship, so reasoning about it is quite focused in scope.

---
cc: @noloco-io/engineering

### CODE REVIEW

File: noloco/utils.py [hunk 1]
```diff
@@ -131,25 +133,42 @@ def find_relationship_data_type(
         # If the relationship field exists on the parent data type then
         # this is a forward relationship and we can simply look up the
         # relationship data type by the corresponding field type.
-        return find_data_type_by_name(relationship_field['type'], data_types)
+        return {
+            'data_type': find_data_type_by_name(
+                relationship_field['type'], data_types),
+            'is_collection': relationship_field['relationship'] == MANY_TO_MANY
```
Comment (darraghmckay) [inline noloco/utils.py:139]: Your `is_collection` calculation isn't quite right 

As you can see from `isMultiRelationship` there's always 2 scenarios that are multi 

```js
export const isMultiRelationship = (relationship) =>
  relationship === ONE_TO_MANY || relationship === MANY_TO_MANY;

export const isReverseMultiRelationship = (relationship) =>
  relationship === MANY_TO_ONE || relationship === MANY_TO_MANY;

export const isMultiField = (field) =>
  (field.relationship && isMultiRelationship(field.relationship)) ||
  (field.relatedField &&
    isReverseMultiRelationship(field.relatedField.relationship));
```

File: noloco/utils.py [hunk 2]
```diff
@@ -131,25 +133,42 @@ def find_relationship_data_type(
...
+                        if reverseName == relationship_name:
+                            return {
+                                'data_type': candidate_data_type,
+                                'is_collection': True
+                            }
+                    # If the reverse name is not populated but this field is a
+                    # one-to-one relationship and the data type name matches
+                    # the relationship name then we have found the data type
+                    # the relationship goes to.
+                    elif field['relationship'] == ONE_TO_ONE:
```
Comment (darraghmckay) [inline noloco/utils.py:166]: As mentioned above, you're omitting the reverse of `ONE_TO_MANY` which is singular, but uses `reverseName` as a key

**Review by darraghmckay — COMMENTED** [review]
While your PR likely fixes the bug found it definitely leaves one or two lying around

I'll leave it up to you whether you want to push the hot-fix

File: noloco/utils.py (hunk 1 above)
Comment (PagingMatt) [reply noloco/utils.py]: Got it, I think that should be corrected now.

**Comment by PagingMatt** [comment]
> While your PR likely fixes the bug found it definitely leaves one or two lying around
> 
> I'll leave it up to you whether you want to push the hot-fix

Would prefer to not leave bugs hanging or worse introduce any new ones so have fixed that calculation.

File: noloco/utils.py [hunk 3]
```diff
@@ -131,25 +135,46 @@ def find_relationship_data_type(
...
+                    # goes to.
+                    if field['reverseName'] is not None and \
+                            field['reverseName'] != '':
+                        reverseName = field['reverseName'] + 'Collection'
+                        if reverseName == relationship_name:
+                            return {
+                                'data_type': candidate_data_type,
+                                'is_collection':
+                                    field['relationship'] == MANY_TO_MANY or
+                                    field['relationship'] == MANY_TO_ONE
```
Comment (darraghmckay) [inline noloco/utils.py:166]: Could make this a util in another PR

**Review by darraghmckay — APPROVED** [review]
Looks great! Nice one 

Do you have to bump the version or anything?

**Comment by PagingMatt** [comment]
> Looks great! Nice one
> 
> Do you have to bump the version or anything?

Yeah, I'll land it and do separately - if we keep it clean, auto generating the version history is easier.

**Comment by darraghmckay** [comment]
> > Looks great! Nice one
> > 
> > Do you have to bump the version or anything?
> 
> Yeah, I'll land it and do separately - if we keep it clean, auto generating the version history is easier.

Ah yes! Forgot about that
